src/D2L.CodeStyle.Analyzers/Async/Generator/AsyncToSyncMethodTransformer.cs

```diff
@@ -0,0 +1,146 @@
+﻿using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace D2L.CodeStyle.Analyzers.Async.Generator;
+
+internal sealed class AsyncToSyncMethodTransformer : SyntaxTransformer {
+	public AsyncToSyncMethodTransformer( SemanticModel model ) : base( model ) { }
+
+	public TransformResult<MethodDeclarationSyntax> Transform( MethodDeclarationSyntax decl ) {
+		// TODO: remove CancellationToken parameters
+		decl = decl.WithAttributeLists( ReplaceGenerateSyncAttribute( decl.AttributeLists ) )
+			.WithModifiers( RemoveAsyncModifier( decl.Modifiers ) )
+			.WithIdentifier( RemoveAsyncSuffix( decl.Identifier ) )
+			.WithReturnType( TransformReturnType( decl.ReturnType ) )
+			.WithExpressionBody( Transform( decl.ExpressionBody ) )
+			.WithBody( Transform( decl.Body ) );
+
+		return GetResult( decl );
+	}
+
+	private SyntaxList<AttributeListSyntax> ReplaceGenerateSyncAttribute(
+		SyntaxList<AttributeListSyntax> attributeLists
+	) => TransformAll( attributeLists, ReplaceGenerateSyncAttribute );
+
+	private AttributeListSyntax? ReplaceGenerateSyncAttribute(
+		AttributeListSyntax attributeList
+	) => attributeList.WithAttributes(
+		TransformAll( attributeList.Attributes, ReplaceGenerateSyncAttribute )
+	);
+
+	private AttributeSyntax ReplaceGenerateSyncAttribute(
+		AttributeSyntax attribute
+	) {
+		// Attributes other than [GenerateSync] pass through
+		if( !IsGenerateSyncAttribute( attribute ) ) {
+			return attribute;
+		}
+
+		// Turn [GenerateSync] into [Blocking].
+		// This makes some assumptions about the ambient using directives.
+		return attribute.WithName(
+			SyntaxFactory.ParseName( "Blocking" ).WithTriviaFrom( attribute.Name )
+		);
+	}
+
+	private SyntaxTokenList RemoveAsyncModifier( SyntaxTokenList modifiers ) =>
+		TransformAll(
+			modifiers,
+			static token => token.IsKind( SyntaxKind.AsyncKeyword ) ? null : token
+		);
+
+	private SyntaxToken RemoveAsyncSuffix( SyntaxToken ident ) {
+		if( !ident.ValueText.EndsWith( "Async" ) || ident.ValueText == "Async" ) {
+			ReportDiagnostic(
+				Diagnostics.ExpectedAsyncSuffix,
+				ident.GetLocation(),
+				ident.ValueText
+			);
+
+			return ident;
+		}
+
+		const int AsyncSuffixLength = 5;
+
+		return SyntaxFactory.Identifier(
+			ident.ValueText.Substring( 0, ident.ValueText.Length - AsyncSuffixLength )
+		).WithTriviaFrom( ident );
+	}
+
+	private TypeSyntax TransformReturnType( TypeSyntax returnType ) {
+		var returnTypeInfo = m_model.GetTypeInfo( returnType );

```

Comment:
Pass CancelationToken to GetTypeInfo

Reply:
It's this running within  the callback to RegisterSourceOutput?

Reply:
No, before that

Reply:
This buddy helps turn an `IncrementalValuesProvider<(MethodDeclarationSyntax, Compilation)>` into an `IncrementalValuesProvider<MethodGenerationResult>` (that type isn't defined yet, but it's the `string` source for one method or diagnostics). From there we `Collect()` them into one big batch and sort them by source file, and then `FileCollector` (previous PR) concatenates them into one file creating an `IncrementalValuesProvider<FileGenerationResult>` (another undefined type -- the `string` for the source of an entire output file + all the associated diagnostics). That part is fed into `RegisterSourceOutput` and the callback there emits the file source (which is a string) and all of the associated diagnostics.